### ISSUE TITLE
Write a script that will create DCPU16 SDK

### ISSUE THREAD

**Issue by krasin** [llvm-dcpu16/llvm-dcpu16#113]
This SDK should include Clang, nos_linker, and probably some emulator.

**Comment by aubreyrjones** [comment]
That's my plan all along. Also with simpler-than-libc library (which is not part of current discussion).

I can do that. You care about scheme? Or is simple git/gem wrapper okay?

**Comment by ghost** [comment]
I would like that your nos_liker becomes a seperate assembler and linker. The reason is simple: CMAKE and other utils have a assembly and a seperate linking stage. If we coul just "replace" the default assembler (gas) and linker (ld) with your programms, we could use a very simple cmake file to create projects.

**Comment by aubreyrjones** [comment]
I don't think I can turn nos_link into a separate assembler and linker. One major problem is that the stage closest to actually assembling is at the very end. I uplift the assembly code into abstract ruby objects quite early, and then manipulate them to link, concatenate, and (eventually) transform the program. Only at the very end, when I have the program "finished", do I fix addresses and assemble.

I _could_ output the abstract token tree as an intermediate file. Try nos_link with '-t' switch to see what that looks like. But, then I raise the objection that it can hardly be called an "assembler" if what it emits is a YAML file 10 times the size and half the readability of the original assembly.

ELF tools have the advantage of a program image binary and an intermediate binary that share nearly identical syntax and semantics. You can load and dynamically "link" single .o files as plugins, if you've got nothing better to do.

DCPU16 does not have the space for a header the size of ELF. In my opinion, it doesn't even have space for a DLL-style symbol list + relocs. I do not believe that I can make a single binary format that is a valid executable as emerges from the assembler, but also carries enough metadata to be linked successfully. This is especially true given that I plan for nos_link to support several high-level features; the amount of metadata necessary for exposing analysis of overlays (and their mutual references), functions, and other multi-instruction constructs is quite substantial.

My other issue is cmake.  I use it in my own development. And it is superior to raw makefiles or .vcproj as build tools go. But, it does have a wickedly steep learning curve. And it brings with it all the baggage of Real World Development. It's also either another external requirement, or another thing we have to package along with the binary release.

One of my goals here is to make an Incredibly Simple system for people to use to build performant DCPU16 applications. We have a very simple system up until now. The CPU is simple. clang is simple to use. nos_link is simple to use. Everything takes a list of files and does its own job. Everything _can_ be configured, but nothing _must_ be configured. But if somebody would want to customize our build process, they would have to learn cmake.

And let's be perfectly honest here, we're not using any of what cmake really offers. Do you imagine us writing cmake library finders for dcpu16 libraries? Or retargeting the project to Visual Studio?

I'd much rather just write a build tool. Let people dump all of their .c files into './project_src', and the build tool does the trivial two commands we need:

```
for_each (File f : ./project_source)
    sfile = clang(f, $CLANG_FLAGS)
end

nos_link(files_in(./build/asm), $LINK_FLAGS)
```

This also has the advantage that you don't need to list all of your project's files in some other file.

**Comment by ghost** [comment]
_mhh_ I see your point. Is you linker right now able to only include those code fragments, that are references by the program. Let's say we write a big standard library, can it sort out the unused stuff right now?

Also: Can you remove the extra dependency for the argument parsing? I'm pretty sure you can do that with the standard ruby library (If we really want to remove the dependencies). We could then build a very easy build system using ruby (since we already got that dependency) and start working on some more "advance" stuff like building an SDK, documenting stuff and start writing a very small libc replacement.

**Comment by aubreyrjones** [comment]
```
Is you linker right now able to only include those code fragments, that are references by the program. Let's say we write a big standard library, can it sort out the unused stuff right now?
```

That is two steps from now. It already detects unused symbols (try -u option). The next step is supporting .func/.endfunc (I will do that tonight, probably). After that, if func symbol is not referenced, then code between .func/.endfunc will not be included. Like, 36 hours from now, I will have that functionality.

```
Can you remove the extra dependency for the argument parsing?
```

That library is _so_ handy and rock solid. I'd prefer not to replicate it.

What I can do, instead, is to include the rubygem as part of whatever we distribute. It is very small, and shipping dependencies is okay in ruby world. We could have one package that has binary clang, trollop gem, nos_link, builder, stdlib, sdk, and all of our custom stuff. Ruby could even be packaged in, if we want. (But it is better to install ruby to the system.)

I only say to go through rubygems program right now because I didn't want to build binary package. But, if binary package is the goal, then we can simply include the trollop rubygem.

**Comment by enkore** [comment]
> That's my plan all along. Also with simpler-than-libc library (which is not part of current discussion).

There is an emerging draft for a minified libc: https://github.com/0x10cStandardsCommittee/0x10c-Standards/issues/35#issuecomment-5152360

**Comment by krasin** [comment]
> Or is simple git/gem wrapper okay?

@netzapper my plan is to have a .tar.gz / zip archive with the SDK that can be extracted to any directory and (eventually) a Debian package for that.

gem is not appropriate, because it makes the process of installation harder (since it adds a dependency on Ruby and rubygems. It's fine to nos_linker, but not fine to Clang)

**Comment by aubreyrjones** [comment]
Well, if the sdk is to include nos_link, it must at least have a dependency on ruby. The trollop gem can be bundled into the package easily, though, removing the rubygems dependency.

I'm not sure how I see the sdk working as a debian installer. I certainly don't want any of this stuff in my real system path. And I can't see any administrator anywhere installing this as a site package.

I had sort of imagined it more like eclipse or rails is now. It's designed to run from local directory. Just unzip and you can start using it, no admin rights required. If you just want dcpu16-clang and nos_link as utilities, you add the 'dcpu16sdk/bin' directory to your $PATH.

If it's shipped as a dev environment, and not just a generic compiler, it could even automate 99% of stuff for kids who're using it. For instance, our build system could have 'dcpu_build new_project mySpiffyProject', and it builds the project scaffolding and everything inside 'dcpu16sdk/projects/mySpiffyProject'.

I think it's _vital_ to remember that our biggest target audience here is brand new programmers. They're going to be directed toward a high-level language, and they're going to greatly appreciate that our toolchain is easier to use than cobbling something together themselves.

Asking them to install ruby, unzip the package, run the scaffold builder, and then they can just start writing code... that's not so bad. It would work even on windows (if clang runs on windows).

**Comment by aubreyrjones** [comment]
Look, here's what I had in mind when I went looking for a C compiler for DCPU16. I will use my name for it, but if you would like to combine efforts, we can call it whatever you like.

```
> cd dcpu/nos
> ./nos new MyProject
> cd projects/MyProject
> emacs include/my_header.h src/my_program.c
(...while I edit files...)
> nos build debug
> nos emulate
(...while I fiddle with it...)
> nos build release
> nos floppy_xfer_webapp     # or however you deploy to in-game
```

So, libraries, compiler, assembler, build tool, tool dependencies, etc. all wrapped up in one directory that somebody can just download and use. And that requires basically 0 configuration on install or on a per-project basis, except when you want non-standard build args or whatever. On Windows, it can even be bundled with ruby and an environment shell launched with a batch file.

(On unix it is better to ask somebody to apt-get ruby, though, I think. But it is now shipped with many distros, and depended on by several common tools.)

The beauty of this is that it's just one (weird) machine target. There is no operating system to link to. And if there is an OS you're linking against, it's going to have to be small and simple. There is no dynamic linking. Projects will not range into the thousands of files, because they literally cannot.

If we build this thing right, not only will a zillion people get into programming because all of the irrelevant technical bullshit was already taken care of... but you and I will get to come home from fighting with continuous integration and the QA department and a Distributed Block Replication system that isn't and doesn't, and have an environment where the technical bullshit doesn't get in the way of having _fun_ programming.

**Comment by ghost** [comment]
So in layman's terms: We want to build a Rails like environment? That's fine with me for now.

**Comment by krasin** [comment]
After thinking a bit, I want to focus on the compiler, because it's still below the bar and let others to create SDKs using Clang.

**Comment by aubreyrjones** [comment]
Well, rails is the runtime too. And does a lot of magic at runtime. 

Maximum build time magic for minimal runtime surprise. Mandatory for operating systems, if you're into that.

**Comment by ghost** [comment]
@krasin You're right, I sure want to focus on other things right now. But we also need a infrastructure to test our work sooner or later with a simple libc and a way to build it.

**Comment by krasin** [comment]
True. But for testing infrastructure I am absolutely fine to have any dependencies, like Java or Ruby or autotools.

**Comment by aubreyrjones** [comment]
So....

I'm working on the debugger. It's coming along okay, conceptually at least. However, packaging is kind of messy.

The basic problem is this: it's hard to write nos_link and nos_db so that they are useful on their own, but when integrated into the SDK suite, have their redundancy stripped. For instance, I use the trollop library in everything. But, if I'm trying not to be dependent on rubygems, I have to include it in each module. That means that instead of 'require rubygems; require trollop', I do 'require_relative ./lib/trollop.rb'. That doesn't work if the whole suite is sharing one copy of trollop, since the path is different.

All things considered, would y'all prefer that the suite itself be approximately zero configuration; or, are y'all not at all sold on my vision, and would prefer to use only bits and pieces in your development (nos_link, but a different emulator, for instance)?

Basically, I'm saying that I need to start working on my environment distribution. If y'all are interested in formally joining forces, then I'm happy to make sure that the SDK suite as a whole is convenient for you. However, if y'all want to go your own way, I will make sure that nos_link always functions as a standalone element.

Either way, I'm going to continue to respond to your requests regarding nos_link. Mainly because I intend to use clang/llvm as the compiler in any case.

[Btw, if we're formally joining up, I'm glad to rename my projects if you don't like the 'nos' prefix.]

**Comment by ghost** [comment]
I wrote a very simple bash script which will take the example files from our "simple-example" repository, llvmp-dcpu16/clang and the binutils-dcpu16 port and will create a small SDK from the scratch. You can find it here:
https://github.com/downloads/llvm-dcpu16/llvm-dcpu16/build-sdk.sh

This should be sufficient for now and I will close this issue.
